### PR TITLE
RichText: fix character loss

### PR DESCRIPTION
**Pull request by WenhaiLin**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/2d/components/rich-text.ts [hunk 1]
```diff
@@ -836,13 +836,13 @@ export class RichText extends Component {
             // concat previous line
             let checkStartIndex = 0;
             while (this._lineOffsetX <= this._maxWidth) {
-                const checkEndIndex = this._getFirstWordLen(labelString, checkStartIndex, labelString.length);
-                const checkString = labelString.substr(checkStartIndex, checkEndIndex);
+                const wordLen = this._getFirstWordLen(labelString, checkStartIndex, labelString.length);
+                const checkString = labelString.substr(checkStartIndex, checkStartIndex + wordLen);
```
Comment (Zhmz) [inline cocos/2d/components/rich-text.ts:840]: image
the second parameter is the string length. why plus the checkStartIndex?

Comment (Zhmz) [reply cocos/2d/components/rich-text.ts:840]: Could you provide an example or a condition about it

Comment (WenhaiLin) [reply cocos/2d/components/rich-text.ts:840]: https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/String/substring

截屏2022-09-08 20 10 35

Comment (WenhaiLin) [reply cocos/2d/components/rich-text.ts:840]: 我们本地改动比较多，已换出substring。等空闲再查是哪一点改动修正的字符丢失

Comment (WenhaiLin) [reply cocos/2d/components/rich-text.ts:840]: 问题：
截屏2022-09-08 20 52 35

参数
```
文本：点点点发行的【收藏卡】将包含线上数字卡以及实体随机卡。我们将线上数字与线下实体结合，用户可根据自己的喜好，在点点点平台， 利用平台内的正版贴
字号：30
行宽：700 
```

触发问题的地方：_updateRichText 函数中调用
```
const splitArr: string[] = this.SplitLongStringApproximatelyIn2048(text, i);
text = splitArr.join('\n');
```

我们本地改了_updateRichText换行的实现，没有使用SplitLongStringApproximatelyIn2048

**Pull request closed without merging**